presto-native-execution/src/test/java/com/facebook/presto/nativeworker/TestHiveWindowQueries.java

```diff
@@ -84,4 +85,105 @@ public void testRowNumber()
     {
         testRankingFunction("row_number()");
     }
+
+    protected List<String> buildWindowTestQueriesWithFrame(List<Pair<String, String>> functionCalls)
+    {
+        ImmutableList.Builder<String> queries = ImmutableList.builder();
+
+        // Window function over clauses with different datatypes in PARTITION BY and ORDER BY clauses.
+        List<String> overClauses = Arrays.asList("PARTITION BY orderkey ORDER BY commitdate asc nulls first, extendedprice asc, shipinstruct desc",
+                "PARTITION BY commitdate ORDER BY orderkey desc, extendedprice asc, shipinstruct desc nulls first",
+                "PARTITION BY extendedprice ORDER BY commitdate asc nulls first, orderkey desc nulls first, shipinstruct asc nulls first",
+                "PARTITION BY shipinstruct ORDER BY commitdate, extendedprice, orderkey",
+                "ORDER BY commitdate desc, orderkey asc, extendedprice desc nulls first, shipinstruct asc nulls first",
+                "ORDER BY orderkey desc, commitdate asc, extendedprice asc nulls first, shipinstruct desc nulls first",
+                "ORDER BY extendedprice asc nulls first, shipinstruct desc nulls first, commitdate desc, orderkey asc",
+                "ORDER BY shipinstruct desc, orderkey asc, commitdate desc nulls first, extendedprice asc nulls first");
+
+        // Window function frame clauses with UNBOUNDED PRECEDING/FOLLOWING and CURRENT ROW bounds in RANGE and ROWS modes.
+        List<String> frameClauses = Arrays.asList("RANGE UNBOUNDED PRECEDING",
+                "RANGE CURRENT ROW",
+                "RANGE BETWEEN UNBOUNDED PRECEDING AND UNBOUNDED FOLLOWING",
+                "RANGE BETWEEN CURRENT ROW AND UNBOUNDED FOLLOWING",
+                "ROWS UNBOUNDED PRECEDING",
+                "ROWS CURRENT ROW",
+                "ROWS BETWEEN UNBOUNDED PRECEDING AND UNBOUNDED FOLLOWING",
+                "ROWS BETWEEN CURRENT ROW AND UNBOUNDED FOLLOWING");
+
+        for (String overClause : overClauses) {
+            for (String frameClause : frameClauses) {
+                StringBuilder windowFunctionCalls = new StringBuilder();
+                for (Pair<String, String> functionCall : functionCalls) {
+                    // Value functions require a total ordering of input rows to produce deterministic results with window frames in ROWS mode. The argument column for the value
+                    // function is therefore passed as a parameter in functionCall and included in the ORDER BY clause.
+                    windowFunctionCalls.append(String.format("%s OVER (%s, %s %s),", functionCall.first(), overClause, functionCall.second(), frameClause));
+                }
+                windowFunctionCalls.deleteCharAt(windowFunctionCalls.length() - 1);
```

Comment:
What is this line for ?

Reply:
This deletes the last ```,``` character after the sequence of window function calls (```%s OVER (%s, %s %s),```).